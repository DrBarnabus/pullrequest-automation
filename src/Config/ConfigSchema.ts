import { z } from 'zod';

export type RequiredApprovals = z.infer<typeof RequiredApprovalsSchema>;
export const RequiredApprovalsSchema = z.union([
  z.number().gte(1),
  z.literal('smart'),
  z
    .object({ baseRef: z.string(), requiredApprovals: z.number().gte(1) })
    .array()
    .nonempty(),
]);

export type LabelsToApply = z.infer<typeof LabelsToApplySchema>;
export const LabelsToApplySchema = z.object({
  approved: z.string(),
  rejected: z.string(),
  needsReview: z.string(),
  draft: z.string().optional(),
});

export type ApprovalLabeller = z.infer<typeof ApprovalLabellerSchema>;
const ApprovalLabellerSchema = z.object({
  enabled: z.literal(true),
  requiredApprovals: RequiredApprovalsSchema,
  useLegacyMethod: z.boolean().optional(),
  labelsToApply: LabelsToApplySchema,
});

export type ApprovalLabellerModule = z.infer<typeof ApprovalLabellerModuleSchema>;
export const ApprovalLabellerModuleSchema = z.discriminatedUnion('enabled', [
  z.object({ enabled: z.literal(false) }),
  ApprovalLabellerSchema,
]);

export type BranchLabeller = z.infer<typeof BranchLabellerSchema>;
export const BranchLabellerSchema = z.discriminatedUnion('enabled', [
  z.object({ enabled: z.literal(false) }),
  z.object({
    enabled: z.literal(true),
    rules: z
      .object({
        baseRef: z.string(),
        headRef: z.string().optional(),
        labelToApply: z.string(),
      })
      .array()
      .nonempty(),
  }),
]);

export type ReviewerExpander = z.infer<typeof ReviewerExpanderSchema>;
export const ReviewerExpanderSchema = z.object({
  enabled: z.boolean(),
});

export type Modules = z.infer<typeof ModulesSchema>;
export const ModulesSchema = z.object({
  approvalLabeller: ApprovalLabellerModuleSchema.optional(),
  branchLabeller: BranchLabellerSchema.optional(),
  reviewerExpander: ReviewerExpanderSchema.optional(),
});

export type CheckForMergeConflictAction = z.infer<typeof CheckForMergeConflictActionSchema>;
export const CheckForMergeConflictActionSchema = z.object({
  action: z.literal('check-for-merge-conflict'),
  name: z.string().optional(),
  with: z.object({
    baseRef: z.string(),
    comparisonBaseRef: z.string(),
  }),
});

export type MergePullRequestAction = z.infer<typeof MergePullRequestActionSchema>;
export const MergePullRequestActionSchema = z.object({
  action: z.literal('merge-pull-request'),
  name: z.string().optional(),
});

export type CreatePullRequestAction = z.infer<typeof CreatePullRequestActionSchema>;
export const CreatePullRequestActionSchema = z.object({
  action: z.literal('create-pull-request'),
  name: z.string().optional(),
  with: z.object({
    baseRef: z.string(),
    headRef: z.string(),
    label: z.string().optional(),
  }),
});

export type Action = z.infer<typeof ActionSchema>;
export const ActionSchema = z.discriminatedUnion('action', [
  CheckForMergeConflictActionSchema,
  MergePullRequestActionSchema,
  CreatePullRequestActionSchema,
]);

export type Command = z.infer<typeof CommandSchema>;
export const CommandSchema = z.object({
  trigger: z.string(),
  actions: ActionSchema.array().nonempty(),
});

export type Config = z.infer<typeof ConfigSchema>;
export const ConfigSchema = z.object({
  modules: ModulesSchema.optional(),
  commands: CommandSchema.array().optional(),
});
