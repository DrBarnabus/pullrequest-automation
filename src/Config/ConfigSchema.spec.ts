import { Config, ConfigSchema } from './ConfigSchema';
import merge from 'lodash/merge';

const baseConfig: Config = {
  modules: {
    approvalLabeller: {
      enabled: true,
      requiredApprovals: 'smart',
      labelsToApply: {
        approved: 'approved',
        rejected: 'rejected',
        needsReview: 'needs-review',
      },
    },
    branchLabeller: {
      enabled: true,
      rules: [
        {
          baseRef: 'development',
          headRef: 'main',
          labelToApply: 'to-main',
        },
      ],
    },
    reviewerExpander: {
      enabled: true,
    },
  },
  commands: [
    {
      trigger: 'Test Command',
      actions: [
        {
          action: 'check-for-merge-conflict',
          with: {
            baseRef: 'development',
            comparisonBaseRef: 'main',
          },
        },
        {
          action: 'merge-pull-request',
          name: 'Merge Pull Request',
        },
        {
          action: 'create-pull-request',
          name: 'Create Pull Request',
          with: {
            baseRef: 'development',
            headRef: 'main',
          },
        },
      ],
    },
  ],
};

describe('ConfigSchema', () => {
  it.each([
    {
      modules: {
        approvalLabeller: { enabled: 'false' },
      },
    },
    {
      modules: {
        approvalLabeller: { requiredApprovals: 0 },
      },
    },
    {
      modules: {
        approvalLabeller: { requiredApprovals: 'invalid' },
      },
    },
    {
      modules: {
        approvalLabeller: { requiredApprovals: {} },
      },
    },
    {
      modules: {
        approvalLabeller: { requiredApprovals: [] },
      },
    },
    {
      modules: {
        approvalLabeller: { requiredApprovals: [{ baseRef: null, requiredApprovals: 0 }] },
      },
    },
    {
      modules: {
        approvalLabeller: { useLegacyMethod: 'false' },
      },
    },
    {
      modules: {
        approvalLabeller: { labelsToApply: null },
      },
    },
    {
      modules: {
        approvalLabeller: {
          labelsToApply: {
            approved: null,
            rejected: null,
            needsReview: null,
          },
        },
      },
    },
    {
      modules: {
        approvalLabeller: {
          labelsToApply: {
            approved: 1,
            rejected: 1,
            needsReview: 1,
            draft: 1,
          },
        },
      },
    },
    {
      modules: {
        branchLabeller: {
          enabled: 'false',
        },
      },
    },
    {
      modules: {
        branchLabeller: {
          rules: null,
        },
      },
    },
    {
      modules: {
        branchLabeller: {
          rules: [],
        },
      },
    },
    {
      modules: {
        branchLabeller: {
          rules: [{ baseRef: 1, headRef: 1, labelToApply: 2 }],
        },
      },
    },
    {
      modules: {
        reviewerExpander: { enabled: 'false' },
      },
    },
    {
      modules: {
        approvalLabeller: 'invalid',
        branchLabeller: 'invalid',
        reviewerExpander: 'invalid',
      },
    },
    {
      commands: {},
    },
    {
      commands: [],
    },
    {
      commands: [
        {
          trigger: 1,
        },
      ],
    },
    {
      commands: [
        {
          actions: {},
        },
      ],
    },
    {
      commands: [
        {
          actions: [],
        },
      ],
    },
    {
      commands: [
        {
          actions: [{ action: 'invalid' }],
        },
      ],
    },
    {
      commands: [
        {
          actions: [{ action: 'check-for-merge-conflict', name: 1 }],
        },
      ],
    },
    {
      commands: [
        {
          actions: [{ action: 'check-for-merge-conflict', with: null }],
        },
      ],
    },
    {
      commands: [
        {
          actions: [{ action: 'check-for-merge-conflict', with: { baseRef: 1, comparisonBaseRef: 1 } }],
        },
      ],
    },
    {
      commands: [
        {
          actions: [{ action: 'merge-pull-request', name: 1 }],
        },
      ],
    },
    {
      commands: [
        {
          actions: [{ action: 'create-pull-request', name: 1 }],
        },
      ],
    },
    {
      commands: [
        {
          actions: [{ action: 'create-pull-request', with: null }],
        },
      ],
    },
    {
      commands: [
        {
          actions: [{ action: 'create-pull-request', with: { baseRef: 1, headRef: 1, label: 1 } }],
        },
      ],
    },
  ])('should correctly flag errors in the configuration %j', (baseAlterationsForTest: unknown) => {
    const objectToTest = merge(baseConfig, baseAlterationsForTest);

    expect(() => ConfigSchema.parse(objectToTest)).toThrowErrorMatchingSnapshot();
  });
});
