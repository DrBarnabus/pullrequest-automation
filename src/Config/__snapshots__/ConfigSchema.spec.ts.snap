// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":[]}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":[{"action":"check-for-merge-conflict","name":1}]}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "name"
    ],
    "message": "Expected string, received number"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":[{"action":"check-for-merge-conflict","with":{"baseRef":1,"comparisonBaseRef":1}}]}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "name"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "with",
      "baseRef"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "with",
      "comparisonBaseRef"
    ],
    "message": "Expected string, received number"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":[{"action":"check-for-merge-conflict","with":null}]}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "name"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "null",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "with"
    ],
    "message": "Expected object, received null"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":[{"action":"create-pull-request","name":1}]}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "name"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "with",
      "baseRef"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "with",
      "headRef"
    ],
    "message": "Required"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":[{"action":"create-pull-request","with":{"baseRef":1,"headRef":1,"label":1}}]}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "name"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "with",
      "baseRef"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "with",
      "headRef"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "with",
      "label"
    ],
    "message": "Expected string, received number"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":[{"action":"create-pull-request","with":null}]}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "name"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "null",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "with"
    ],
    "message": "Expected object, received null"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":[{"action":"invalid"}]}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_union_discriminator",
    "options": [
      "check-for-merge-conflict",
      "merge-pull-request",
      "create-pull-request"
    ],
    "path": [
      "commands",
      0,
      "actions",
      0,
      "action"
    ],
    "message": "Invalid discriminator value. Expected 'check-for-merge-conflict' | 'merge-pull-request' | 'create-pull-request'"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":[{"action":"merge-pull-request","name":1}]}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "actions",
      0,
      "name"
    ],
    "message": "Expected string, received number"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"actions":{}}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":[{"trigger":1}]} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "commands",
      0,
      "trigger"
    ],
    "message": "Expected string, received number"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"commands":{}} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":"invalid","branchLabeller":"invalid","reviewerExpander":"invalid"}} 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "approvalLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "branchLabeller"
    ],
    "message": "Expected object, received string"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander"
    ],
    "message": "Expected object, received string"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"enabled":"false"}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"labelsToApply":{"approved":1,"rejected":1,"needsReview":1,"draft":1}}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"labelsToApply":{"approved":null,"rejected":null,"needsReview":null}}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"labelsToApply":null}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"requiredApprovals":"invalid"}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"requiredApprovals":[]}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"requiredApprovals":[{"baseRef":null,"requiredApprovals":0}]}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"requiredApprovals":{}}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"requiredApprovals":0}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"approvalLabeller":{"useLegacyMethod":"false"}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"branchLabeller":{"enabled":"false"}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  },
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "branchLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"branchLabeller":{"rules":[]}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  },
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "branchLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"branchLabeller":{"rules":[{"baseRef":1,"headRef":1,"labelToApply":2}]}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  },
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "branchLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"branchLabeller":{"rules":null}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  },
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "branchLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  }
]"
`;

exports[`ConfigSchema should correctly flag errors in the configuration {"modules":{"reviewerExpander":{"enabled":"false"}}} 1`] = `
"[
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "approvalLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  },
  {
    "code": "invalid_union_discriminator",
    "options": [
      false,
      true
    ],
    "path": [
      "modules",
      "branchLabeller",
      "enabled"
    ],
    "message": "Invalid discriminator value. Expected false | true"
  },
  {
    "code": "invalid_type",
    "expected": "boolean",
    "received": "string",
    "path": [
      "modules",
      "reviewerExpander",
      "enabled"
    ],
    "message": "Expected boolean, received string"
  }
]"
`;
